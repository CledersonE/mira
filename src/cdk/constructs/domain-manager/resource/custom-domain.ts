import { Topic } from '@aws-cdk/aws-sns'
import { Construct } from '@aws-cdk/core'
import {
  CustomResource,
  CustomResourceProvider
} from '@aws-cdk/aws-cloudformation'
import { MiraConfig } from '../../../..'
import { DOMAIN_SUBSCRIPTION_TOPIC } from './route53-manager'

export interface CustomDomainProps {
  source: string
  target: string
}

export class CustomDomain extends Construct {
  domain: CustomResource

  constructor (parent: Construct, props: CustomDomainProps) {
    super(parent, 'CustomDomainResouce')

    const topicName = MiraConfig.getBaseStackName(DOMAIN_SUBSCRIPTION_TOPIC)
    const domainAccountConfig = MiraConfig.getEnvironment('domain')

    const domainProviderTopic = Topic.fromTopicArn(
      this,
      'DomainProvider',
      `arn:aws:sns:${domainAccountConfig.env.region}:${domainAccountConfig.env.account}:${topicName}`
    )

    this.domain = new CustomResource(this, 'CustomDomain', {
      provider: CustomResourceProvider.fromTopic(domainProviderTopic),
      properties: {
        Source: props.source,
        Target: props.target
      },
      resourceType: 'Custom::Domain'
    })
  }
}
