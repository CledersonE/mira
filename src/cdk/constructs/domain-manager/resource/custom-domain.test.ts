import { SynthUtils } from "@aws-cdk/assert"
import { Stack, App } from "@aws-cdk/core"

import {
  MiraConfig,
  DomainConfig,
  Account,
} from "../../../../config/mira-config"

import { CustomDomain } from "./custom-domain"

describe("CustomDomain", () => {
  test("renders output correctly", () => {
    const stack = new Stack(
      new App(),
      MiraConfig.getBaseStackName("CustomDomain"),
      {}
    )
    MiraConfig.getEnvironment = (envName: string): Account => {
      if (envName === 'domain') {
        return {
          name: "domain",
          profile: "domain",
          env: { account: "domain", region: "eu-west-1" },
        }
      }

      return {
        name: "account1",
        profile: "account1",
        env: { account: "account1", region: "eu-west-1" },
      }
    }

    MiraConfig.getDomainConfig = (): DomainConfig => ({
      hostedZoneId: 'Z123456789',
      accounts: [
        'account1',
        'domain'
      ],
    })

    const f = new CustomDomain(stack, {
      source: 'source-url',
      target: 'target-url'
    })
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
  })
})
