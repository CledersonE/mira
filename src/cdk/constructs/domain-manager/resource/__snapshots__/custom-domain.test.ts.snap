// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomDomain renders output correctly 1`] = `
Object {
  "Resources": Object {
    "CustomDomainResourceCustomDomain51E2934C": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": "arn:aws:sns:eu-west-1:domain:John-MyGreatApp-DomainSubscriptionTopic",
        "Source": "source-url",
        "Target": "target-url",
      },
      "Type": "Custom::Domain",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
